xray/grpc.go

```diff
@@ -0,0 +1,174 @@
+		return Capture(ctx, host, func(ctx context.Context) error {
+			seg := GetSegment(ctx)
+			if seg == nil {
+				return errors.New("failed to record gRPC transaction: segment cannot be found")
+			}
+
+			ct, e := NewClientTrace(ctx)
+			if e != nil {
+				return e
+			}
+			md := metadata.Pairs(TraceIDHeaderKey, seg.DownstreamHeader().String())
+			ctx2 := metadata.NewOutgoingContext(httptrace.WithClientTrace(ctx, ct.httpTrace), md)
+
+			seg.Lock()
+			seg.Namespace = "remote"
+			seg.GetHTTP().GetRequest().URL = "grpc://" + host + method
+			seg.GetHTTP().GetRequest().Method = method
+			seg.Unlock()
+
+			err := invoker(ctx2, method, req, reply, cc, opts...)
+
+			recordContentLength(seg, reply)
+			if err != nil {
+				classifyErrorStatus(seg, err)
+				ct.subsegments.GotConn(nil, err)
+			}
+
+			return err
+		})
+	}
+}
+
+func unaryServerInterceptorWithConfig(sn SegmentNamer, cfg *Config) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+
+		var traceID string
+		if ok && len(md.Get(TraceIDHeaderKey)) == 1 {
+			traceID = md.Get(TraceIDHeaderKey)[0]
+		}
+		traceHeader := header.FromString(traceID)
+
+		var host string
+
+		if len(md.Get(":authority")) == 1 {
+			host = md.Get(":authority")[0]
+		}
+		requestURL := url.URL{
+			Scheme: "grpc",
+			Host:   host,
+			Path:   info.FullMethod,
+		}
+		name := sn.Name(host)
+
+		ctx = context.WithValue(ctx, RecorderContextKey{}, cfg)
+
+		var seg *Segment
+		ctx, seg = NewSegmentFromHeader(ctx, name, &http.Request{
+			Host:   host,
+			URL:    &requestURL,
+			Method: info.FullMethod,
+		}, traceHeader)
+		defer seg.Close(nil)
+
+		seg.Lock()
+		seg.GetHTTP().GetRequest().ClientIP, seg.GetHTTP().GetRequest().XForwardedFor = clientIPFromGrpcMetadata(md)
+		seg.GetHTTP().GetRequest().URL = requestURL.String()
+		seg.GetHTTP().GetRequest().Method = info.FullMethod
+		if len(md.Get("user-agent")) == 1 {
+			seg.GetHTTP().GetRequest().UserAgent = md.Get("user-agent")[0]
+		}
+		seg.Unlock()
+
+		resp, err = handler(ctx, req)
+		if err != nil {
+			classifyErrorStatus(seg, err)
+		}
+		recordContentLength(seg, resp)
+		if headerErr := addResponseTraceHeader(ctx, seg, traceHeader); headerErr != nil {
+			err = multierr.Combine(err, headerErr)
+		}
+
+		return resp, err
+	}
+}
+
+// UnaryServerInterceptorWithContext provides gRPC unary server interceptor. xray.Config of given ctx will be inherited
+// by the instrumented context.
+func UnaryServerInterceptorWithContext(ctx context.Context, sn SegmentNamer) grpc.UnaryServerInterceptor {
+	cfg := GetRecorder(ctx)
+	return unaryServerInterceptorWithConfig(sn, cfg)
+}
+
+// UnaryServerInterceptor provides gRPC unary server interceptor.
+func UnaryServerInterceptor(sn SegmentNamer) grpc.UnaryServerInterceptor {
```

Comment:
It would be nice to have a default interceptor that uses the gRPC service name as the segment name

Reply:
Agree. So I guess there will be four variants of the server interceptor?
1. `UnaryServerInterceptor()`
1. `UnaryServerInterceptorWithContext(ctx context.Context)`
1. `UnaryServerInterceptorWithNamer(sn SegmentNamer)`
1. `UnaryServerInterceptorWithContextAndNamer(ctx context.Context, sn SegmentNamer)`

Reply:
Yeah - can you use the options pattern to have some symmetry with upstream gRPC and avoid the need for the explosion of factory methods?

https://github.com/grpc/grpc-go/blob/master/server.go#L118

Reply:
Fixed